webapp/src/cloud.ts

```diff
@@ -0,0 +1,298 @@
+            delete remoteHeaderMap[local.id];
+            if (remote) {
+                local.cloudLastSyncTime = remote.cloudLastSyncTime
+                // Note that we use modification time to detect differences. If we had full (or partial) history, we could
+                //  use version numbers. However we cannot currently use etags since the Cosmos list operations
+                //  don't return etags per-version. And because of how etags work, the record itself can never
+                //  have the latest etag version.
+                if (local.modificationTime !== remote.modificationTime) {
+                    const remoteFile = await getAsync(local);
+                    if (local.cloudCurrent) {
+                        // No local changes, download latest.
+                        const newHeader = {...local, ...remoteFile.header} // make sure we keep local-only metadata like _rev
+                        localHeaderChanges[remoteFile.header.id] = newHeader
+                        await workspace.saveAsync(newHeader, remoteFile.text, true);
+                    } else {
+                        // Possible conflict.
+                        const conflictStr = `conflict found for '${local.name}' (${local.id.substr(0, 5)}...). Last cloud change was ${agoStr(remoteFile.header.modificationTime)} and last local change was ${agoStr(local.modificationTime)}.`
+                        // last write wins.
+                        if (local.modificationTime > remoteFile.header.modificationTime) {
+                            if (local.cloudVersion === remoteFile.version) {
+                                // local is one ahead, push as normal
+                                pxt.debug(`local project '${local.name}' has changes that will be pushed to the cloud.`)
+                            } else {
+                                // conflict and local wins
+                                // TODO: Pop a dialog and/or show the user a diff. Better yet, handle merges.
+                                pxt.log(conflictStr + ' Local will overwrite cloud.')
+                            }
+                            const text = await workspace.getTextAsync(local.id);
+                            const newVer = await setAsync(local, remoteFile.version, text);
+                            U.assert(!!newVer, 'Failed to sync local change (1)');
+                            // save to the workspace header again to make sure cloud metadata gets saved
+                            await workspace.saveAsync(local, null, true)
+                        } else {
+                            // conflict and remote wins
+                            // TODO: Pop a dialog and/or show the user a diff. Better yet, handle merges.
+                            pxt.log(conflictStr + ' Cloud will overwrite local.')
+                            const newHeader = {...local, ...remoteFile.header} // make sure we keep local-only metadata like _rev
```

Comment:
One code clean up item I'd like to do in a subsequent PR is to create "transferToCloud" "transferFromCloud" functions because there is a lot of duplicate work here